Integrated tests with Lerna (does not include Block for now)

evertonfraga opened the pull request:
Down the road of the monorepo transition, this PR introduces integrated module testing, powered by `Lerna bootstrap` command, with some additional configuration. This enables us to detect possible disruption of downstream libraries before releasing.

The monorepo codebase had Block in a non-production state (3.0.0), that wasn't integrated to the other packages at all, making this very transition a bit cumbersome. 


Note to reviewers: I advise review looking at one commit at a time. It contains:

1. Reverts Block code to 2.2.2. This is an inflated commit ([12d503d](https://github.com/ethereumjs/ethereumjs-vm/commit/12d503ded0ac3035c1949751fd9cfbc215b1823d))
1. Removes NodeJS 8.x from tests (soon we'll include node 14) ([e8e3201](https://github.com/ethereumjs/ethereumjs-vm/commit/e8e3201e719ecf65065cea38e696f20dfe17eff0))
1. Removes specific code and scripts for running tests with ts-node with linked packages, which will come at a later stage. ([74890a2](https://github.com/ethereumjs/ethereumjs-vm/commit/74890a211c125883dae7582c2f6cbb2f162cdf7e)) and  ([8d5f3c0](https://github.com/ethereumjs/ethereumjs-vm/commit/8d5f3c0532fd2b4c875fddff1ca0b8672583fc34))
1. Configures test coverage using the same standard as other ethereumjs-VM packages([ca04215](https://github.com/ethereumjs/ethereumjs-vm/commit/ca042152ae2cc962cc7aa82c14b3d31da420cb97))
1. Puts `lerna` commands in place ([75b0479](https://github.com/ethereumjs/ethereumjs-vm/commit/75b0479d2701e036e1c5481fd7179df56103cc45)).

This commit ([75b0479](https://github.com/ethereumjs/ethereumjs-vm/commit/75b0479d2701e036e1c5481fd7179df56103cc45)) also disables `node_modules` cache on GitHub Actions test suite. They consider shallow paths (`./node_modules`), and generic caching keys, which was 100% correct before the monorepo transition. Related: https://github.com/ethereumjs/ethereumjs-vm/issues/730

This PR replaces #725 and should be merged to `master` instead.

holgerd77 commented:
This "revert Block code to 2.2.2" thing makes me a bit nervous since it is a pretty heavy and invasive step, is this really necessary to achieve what you want? How will you achieve to get these changes in again with a retained authorship and stuff like that? 🤔

evertonfraga commented:
You were right @holgerd77, and in the end I just wanted to verify that all the production packages worked well in this new scenario, no need for it to reach `master`. It did worked nicely, then I moved back to Block v3 source code.

After a surgical rebase, I removed the `Revert Block to 2.2.2`, along with some other Block-specific commits. I also edited the `test: enabling Lerna for integrated tests`, excluding the Lerna changes to Block. I have amended the commit message to `test: enabling Lerna for integrated tests, all packages but Block`.

These changes guarantees all packages (but Block ) work well in conjunction, , and can finally tackle the Block v3 migration with more confidence.

evertonfraga commented on .circleci/config.yml:
```diff
@@ -25,7 +28,19 @@ jobs:
+      - checkout
+      - run: 
+          name: Installing dependencies
+          command: cd ~/project/packages/${CIRCLE_JOB} && npm install
```
For this PR, `Block` won't use Lerna for linking packages. So in both CIs, we need to explicitly `npm install` for `Block` package.

evertonfraga commented on .github/workflows/vm-nightly-test.yml:
```diff
@@ -3,64 +3,64 @@ name: VM Nightly Test
+    #       CI: true
+    #     working-directory: '${{ env.cwd }}'
+    # env:
+    #   cwd: '${{github.workspace}}/packages/vm'
```
After migrating to a monorepo setup, these cache warm-up jobs are quite useless, as they only reach the shallow `node_modules`. We should reactivate them later, probably along with Lerna package hoisting.

evertonfraga commented on lerna.json:
```diff
@@ -2,5 +2,10 @@
+  "version": "0.0.0",
+  "command": {
+    "bootstrap": {
+      "ignore": "ethereumjs-block"
```
This is important: it tells Lerna to ignore `ethereumjs-block` from `lerna bootstrap` commands. It will still be used on `lerna run` and `lerna exec` for the time being.

ryanio commented on packages/tx/.prettierignore:
```diff
@@ -2,4 +2,5 @@ node_modules
-.nyc_output
\ No newline at end of file
+.nyc_output
+docs
```
Great, would love to add this to our global prettier config in the future 👍

holgerd77 commented:
👍
